engine/common/synchronization/engine.go

```diff
@@ -269,23 +158,28 @@ func (e *Engine) setupResponseMessageHandler() error {
 	return nil
 }
 
-// Ready returns a ready channel that is closed once the engine has fully
-// started. For consensus engine, this is true once the underlying consensus
-// algorithm has started.
+// Ready returns a ready channel that is closed once the engine has fully started.
 func (e *Engine) Ready() <-chan struct{} {
-	e.unit.Launch(e.checkLoop)
-	for i := 0; i < defaultEngineRequestsWorkers; i++ {
-		e.unit.Launch(e.requestProcessingLoop)
-	}
-	e.unit.Launch(e.responseProcessingLoop)
-	e.unit.Launch(e.finalizationProcessingLoop)
-	return e.unit.Ready()
+	e.lm.OnStart(func() {
```

Comment:
what is the benefit of using lifetime manager here?

Reply:
Concurrency-safety and idempotency.

Without it, if I call `Ready()` multiple times in a row, or I call it concurrently, the loops will be launched multiple times.

Please see https://github.com/dapperlabs/flow-go/issues/5702 and https://github.com/onflow/flow-go/blob/master/module/common.go#L3-L17 for context.

Reply:
> Is there specific case where this rule doesn't apply anymore?

Yes there are. For example, in the [splitter network](https://github.com/onflow/flow-go/blob/smnzhu/sync-request-handler/engine/common/splitter/network/network.go#L112), we must wait for the underlying network to be ready before we are ready. However, the underlying network's `Ready` method is also called by the NodeBuilder logic when starting up all the components. 

More generally though, even if the owner of a component is the only one that manages its lifecycle, how can you guarantee that the owner calls `Ready` and `Done` at most once? If you want to guarantee this in the face of concurrency, then you will need to add synchronization logic in the owner's code, and as a result you've merely moved the problem a bit higher up the stack. It is much better imo to have a standardized way of dealing with concurrency-safety, and having well-defined conventions means that we have less things to worry about in terms of managing child components.

Reply:
See https://github.com/onflow/flow-go/pull/1077#issuecomment-892848052 also

Reply:
Yeah, I agree the API would be more clear if we have a `Run` to start the engine, and a `Ready` function to return the channel indicating "is ready" signal.

Right now. The `Ready` function is the `Run` function, and the returned channel is the `is ready` signal. 

So the current rule is that:
If engine A owns engine B, and manages B's lifecycle, then in A's `Ready`, it will wait for `B`'s `Ready`. and `A`'s `Done` will wait for `B`'s `Done`. This is the case for [the compliance engine (A) manages hotstuff engine (B)](https://github.com/onflow/flow-go/blob/master/engine/consensus/compliance/engine.go#L168)

Now our new scenario is that engine A owns engine B, and engine C wants to wait until engine B is ready. 
If we have the `Run` and `Ready` methods separated, then engine A's `Run` method could call engine B's `Run` method, and engine A's Ready will wait for engine B's Ready. 
And if engine C takes engine B as constructor argument, then engine C's Ready could wait for engine B's Ready.
The reason engine A owns engine B is that when engine A is Stopping, it will also stop engine B (managing its lifecycle). Whereas when engine C is stopping, it will not stop engine B.

IMO, this would be a more clear way than involving concurrency. If we make the ownership clear, then there will be no concurrency issue on calling the `Run` method. An engine's Run method should just be called once. The Ready method can be called concurrently, since it's just returning a read-only channel.

Reply:
Yep, glad you agree. For now though, we can keep the lifecycleManager imo because there are existing cases in the codebase where `Ready` will be called multiple times. Even in this PR: both the request handler and the response handler engines should wait for the `FinalizedHeaderCache`'s `Ready`.

I'll summarize everything into an issue so we can come back to this.